jax/experimental/pjit.py

```diff
@@ -0,0 +1,211 @@
+
+  out_nodes = xla.jaxpr_subcomp(
+      subc, call_jaxpr, backend, axis_env, (),
+      extend_name_stack(name_stack, wrap_name(name, "pjit")), *args)
+  out_mesh_axes = out_mesh_axes_thunk()
+  assert len(out_mesh_axes) == len(out_nodes)
+  out_nodes = [xb.set_sharding_proto(subc, out,
+                                     get_sharding_proto(c, n, mesh_axes, mesh))
+               for out, mesh_axes in safe_zip(out_nodes, out_mesh_axes)]
+
+  subc = subc.build(xops.Tuple(subc, out_nodes))
+  return xops.Call(c, subc, list(in_nodes))
+
+pjit_call_p = core.CallPrimitive("pjit_call")
+pjit_call_p.def_impl(_pjit_call_impl)
+xla.call_translations[pjit_call_p] = _pjit_translation_rule
+
+# None indicates replicated axis (partial replication)
+ArrayAxisPartitioning = Union[pxla.MeshAxisName, Tuple[pxla.MeshAxisName, ...], None]
+# None indicates fully replicated array value
+ArrayPartitioning = Optional[Tuple[ArrayAxisPartitioning, ...]]
+
+@lu.cache
+def _pjit_callable(
+    fun: lu.WrappedFun,
+    in_mesh_axes: Tuple[ArrayPartitioning, ...],
+    out_mesh_axes_thunk: Callable[[], Tuple[ArrayPartitioning, ...]],
+    resource_env,
+    name: str,
+    *in_avals):
+
+  in_axes = [get_array_mapping(axes) for axes in in_mesh_axes]
+  out_axes_thunk = lambda: [get_array_mapping(axes) for axes in out_mesh_axes_thunk()]
+  # TODO(skye): allow for using a submesh of physical_mesh
+  return pxla.mesh_tiled_callable(fun, name, None, resource_env.physical_mesh,
+                                  in_axes, out_axes_thunk, True, *in_avals,
+                                  fun_includes_mesh_axes=True)
+
+
+def with_sharding_constraint(x, mesh_axes):
+  # TODO(skye): this should arguably be part of axis_env
+  resource_env = maps.thread_resources.env
+  return sharding_constraint_p.bind(x, mesh_axes=mesh_axes,
+                                    resource_env=resource_env)
+
+def _sharding_constraint_impl(x, mesh_axes, resource_env):
+  # TODO(skye): can we also prevent this from being called in other
+  # non-pjit contexts? (e.g. pmap, control flow)
+  raise NotImplementedError(
+      "with_sharding_constraint() should only be called inside pjit()")
+
+def _sharding_constraint_translation_rule(c, x_node, mesh_axes, resource_env):
+  mesh = resource_env.physical_mesh
+  return xb.set_sharding_proto(c, x_node,
+                               get_sharding_proto(c, x_node, mesh_axes, mesh))
+
+sharding_constraint_p = core.Primitive("sharding_constraint")
+sharding_constraint_p.def_impl(_sharding_constraint_impl)
+sharding_constraint_p.def_abstract_eval(lambda x, **unused: x)
+ad.deflinear2(sharding_constraint_p,
+              lambda ct, _, mesh_axes, resource_env: (
+                  with_sharding_constraint(ct, mesh_axes),))
+xla.translations[sharding_constraint_p] = _sharding_constraint_translation_rule
+
+
+def get_array_mapping(mesh_axes: ArrayPartitioning) -> pxla.ArrayMapping:
```

Comment:
This gets called all over the place. Why can't we just use the same API as in xmap and avoid converting back and forth between partition specs and the dicts?

Reply:
After talking to Adam, we concluded it should be pretty easy to use the xmap API and get the best of all worlds. I'm gonna do this in a follow-up to get this in while I figure out how xmap does it.